pkg/image/merge.go

```diff
@@ -0,0 +1,185 @@
+	"path"
+	"strings"
+	"text/template"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/labring/sealos/pkg/constants"
+	"github.com/labring/sealos/pkg/image/types"
+	"github.com/labring/sealos/pkg/types/v1beta1"
+	fileutil "github.com/labring/sealos/pkg/utils/file"
+	"github.com/labring/sealos/pkg/utils/logger"
+	"github.com/labring/sealos/pkg/utils/maps"
+	"github.com/labring/sealos/pkg/utils/rand"
+)
+
+func Merge(newImageName string, images []string, buildOptions *types.BuildOptions) error {
+	registrySvc, err := NewRegistryService()
+	if err != nil {
+		return err
+	}
+
+	clusterSvc, err := NewClusterService()
+	if err != nil {
+		return err
+	}
+	imageSvc, err := NewImageService()
+	if err != nil {
+		return err
+	}
+	mergeDir, err := fileutil.MkTmpdir("")
+	if err != nil {
+		return err
+	}
+	logger.Debug("mergeDir: %s", mergeDir)
+	var containerNames []string
+	eg, _ := errgroup.WithContext(context.Background())
+	for _, v := range images {
+		imageName := strings.TrimSpace(v)
+		if imageName == "" {
+			continue
+		}
+		eg.Go(func() error {
+			err = registrySvc.Pull(imageName)
+			if err != nil {
+				return err
+			}
+			cName := fmt.Sprintf("%s-%s", imageName, rand.Generator(8))
+			var manifest *types.ClusterManifest
+			if manifest, err = clusterSvc.Create(cName, imageName); err != nil {
+				logger.Warn("container create failed: %+v", err)
+				return err
+			}
+			containerNames = append(containerNames, manifest.Container)
+			_ = fileutil.RecursionCopy(manifest.MountPoint, mergeDir)
+			return nil
+		})
+	}
+	if err = eg.Wait(); err != nil {
+		return err
+	}
+
+	defer func() {
+		for _, i := range containerNames {
+			_ = clusterSvc.Delete(i)
+		}
+		_ = os.RemoveAll(mergeDir)
+	}()
+
+	ociList, err := imageSvc.Inspect(images...)
+	if err != nil {
+		return err
+	}
+	dockerfile, err := mergeMountPath(ociList, images...)
+	if err != nil {
+		return err
+	}
+	err = os.WriteFile(path.Join(mergeDir, "Kubefile"), []byte(dockerfile), 0755)
+	if err != nil {
+		return err
+	}
+	buildOptions.File = path.Join(mergeDir, "Kubefile")
+	buildOptions.Offline = true
+	buildOptions.Tag = newImageName
+	_ = os.RemoveAll(path.Join(mergeDir, "images", "shim"))
+	logger.Debug("buildOptions file: %s", buildOptions.File)
+	logger.Debug("buildOptions tag: %s", newImageName)
+	logger.Debug("buildOptions contextDir: %s", mergeDir)
+	return imageSvc.Build(buildOptions, mergeDir, buildOptions.Tag)
+}
+func mergeMountPath(ociList types.ImageListOCIV1, images ...string) (string, error) {
+	var labels map[string]string
+	var envs map[string]string
+	var cmds []string
+	var entrypoints []string
+	var isRootfs bool
+	for i, oci := range ociList {
+		labels = maps.MergeMap(labels, oci.Config.Labels)
+		if oci.Config.Labels != nil && oci.Config.Labels[constants.ImageTypeKey] == string(v1beta1.RootfsImage) {
+			isRootfs = true
+		}
+		envs = maps.MergeMap(envs, maps.ListToMap(oci.Config.Env))
+		cmds = append(cmds, oci.Config.Cmd...)
+		if i == len(ociList)-1 {
```

Comment:
@fanux It seems that Entrypoint doesn't have a better solution to merge?

Reply:
why not same with CMD?